man/pkgadd.8.in
.TH pkgadd 8 "" "cards #VERSION#" ""
.SH NAME
pkgadd \- install software package
.SH SYNOPSIS
\fBpkgadd [options] <file>\fP
.SH DESCRIPTION
\fBpkgadd\fP is a \fIpackage management\fP utility, which installs
a software package. A \fIpackage\fP is an archive of files (.pkg.tar.xz).
.SH OPTIONS
.TP
.B "\-i, \-\-ignore"
Ignore the pre and post install of the package. Needed for the install script.
.TP
.B "\-u, \-\-upgrade"
Upgrade/replace package with the same name as <file>.
.TP
.B "\-f, \-\-force"
Force installation, overwrite conflicting files. If the package
that is about to be installed contains files that are already
installed this option will cause all those files to be overwritten.
This option should be used with care, preferably not at all.
.TP
.B "\-r, \-\-root <path>"
Specify alternative installation root (default is "/"). This
should \fInot\fP be used as a way to install software into
e.g. /usr/local instead of /usr. Instead this should be used
if you want to install a package on a temporary mounted partition,
which is "owned" by another system. By using this option you not only
specify where the software should be installed, but you also
specify which package database to use.
.TP
.B "\-v, \-\-version"
Print version and exit.
.TP
.B "\-h, \-\-help"
Print help and exit.
.SH CONFIGURATION
\fBpkgadd\fP is configured by the file \fI/etc/pkgadd.conf\fP. This file can contain rules, that are built out of three fragments: \fIevent\fP, \fIpattern\fP and \fIaction\fP. The event describes in what kind of situation this rule applies. Currently there are two types of events: \fBUPGRADE\fP and \fBINSTALL\fP. \fBUPGRADE\fP rules are applied when a package is installed over an existing version, and \fBINSTALL\fP rules are applied in any case. The pattern is a regular expression. The action applicable to both the \fBUPGRADE\fP and \fBINSTALL\fP event is \fBYES\fP and \fBNO\fP. More than one rule of the same
event type is allowed, in which case the first rule will have the lowest priority and the last rule
will have the highest priority. Example:

.nf
UPGRADE         ^etc/.*$                NO
UPGRADE         ^var/log/.*$            NO
UPGRADE         ^etc/X11/.*$            YES
UPGRADE         ^etc/X11/XF86Config$    NO
.fi

The above example will cause pkgadd to never upgrade anything in /etc/ or /var/log/ (subdirectories included),
except files in /etc/X11/ (subdirectories included), unless it is the file /etc/X11/XF86Config.
The default rule is to upgrade/install everything, rules in this file are exceptions to that rule.
(NOTE! A \fIpattern\fP should never contain an initial "/" since you are referring to the files in the
package, not the files on the disk.)

If pkgadd finds that a specific file should not be upgraded it will install it under \fI/var/lib/pkg/rejected/\fP.
The user is then free to examine/use/remove that file manually.
.SH FILES
.TP
.B "/etc/pkgadd.conf"
Configuration file.
.SH SEE ALSO
pkgrm(8), pkginfo(8), pkgmk(8), rejmerge(8)
.SH COPYRIGHT
This program is part of the *p_cards package* Copyright (c) 2016, **peter1000** <https://github.com/peter1000>  
which is licensed under the GNU General Public License.
*p_cards* includes code derived from other open source projects, which have their own licenses:
Read the LICENSE.md file for the detailed license information.


CHANGELOG.md
# v1.99.9 (xxxx-xx-xx)

## Summary

* Renamed to `p_cards` peter1000 development repo
* Updated: CHANGELOG.md, LICENSE.md, README.md, TODO.md
* Adds: initial documentation structure for [MkDocs](https://github.com/mkdocs/mkdocs)
* Removes french manpages
* Removes sub-package *fltk*
* Replaces `pkgmk.in` with `pkgmk2.in`
* Removes individual license headers in favour of the overall package license.
* Adds *Christophe Devine*, *Junio C Hamano* and *Logan Ingalls* to the Copyright holder in LICENSE.md

# History

* On 07. March 2016 **peter1000** <https://github.com/peter1000/> renamed 
    `cards commit 89580529259f5e912a2efc936bebf90e3f698537` to **p_cards**.

* [cards-1.99.8](https://github.com/NuTyX/cards/commit/992a2ca11bf771e0809ebe9d0987bdc1d47a8e82)    (on Mar 03, 2016)
* [cards-1.99.0](https://github.com/NuTyX/cards/commit/d9d4ae224175ab8514b1951a0078cc0fb75d60fc)    (on Jan 14, 2016)
* [cards-1.96.0](https://github.com/NuTyX/cards/commit/ae6c74e2f2f463bea0699ab2bb43f5daa3df15f2)    (on Jan 03, 2016)
* [cards-1.95.0](https://github.com/NuTyX/cards/commit/9f37977d754bd1dd88346f28e50ea629a69034d8)    (on Jan 02, 2016)
* [cards-1.4.0](https://github.com/NuTyX/cards/commit/f7bd5a8d341eb63aa0a772a5d8e3c201d2e97b15)     (on Dec 25, 2015)
* [cards-1.3.0](https://github.com/NuTyX/cards/commit/a695725e37beb64bda56ca000c5ee86d1f34541e)     (on Aug 25, 2015)
* [cards-1.2.1](https://github.com/NuTyX/cards/commit/14b580fc4165f256e952262914d9cb7e2cba6f74)     (on Jul 23, 2015)
* [cards-1.1](https://github.com/NuTyX/cards/commit/f3eb1243dabec33f7854dd6330a53438d62c8440)       (on Jun 06, 2015)
* [cards-1.0](https://github.com/NuTyX/cards/commit/49452594a9fc8c312fdc4c7203d76761ec5eab9f)       (on May 14, 2015) 


src/pkgadd.cpp
#include "file_utils.h"
#include "pkgadd.h"
#include "process.h"

#include <algorithm>

#include <fstream>
#include <iterator>
#include <cstdio>
#include <regex.h>
#include <unistd.h>

Pkgadd::Pkgadd()
	: Pkgdbh("pkgadd"),
	m_runPrePost(true),
	m_upgrade(false),
	m_force(false)
{
}
Pkgadd::Pkgadd(const string& commandName)
	: Pkgdbh(commandName),
	m_runPrePost(true),
	m_upgrade(false),
	m_force(false)
{
}
void Pkgadd::run(int argc, char** argv)
{
	parseArguments(argc, argv);
	run();
	finish();
}
void Pkgadd::parseArguments(int argc, char** argv)
{
	for (int i = 1; i < argc; i++) {
		string option(argv[i]);
		if (option == "-r" || option == "--root") {
			assertArgument(argv, argc, i);
			m_root = argv[i + 1];
			i++;
		} else if (option == "-i" || option == "--ignore") {
			m_runPrePost = false;
		} else if (option == "-u" || option == "--upgrade") {
			m_upgrade = true;
		} else if (option == "-f" || option == "--force") {
			m_force = true;
		} else if (option[0] == '-' || !m_packageArchiveName.empty()) {
			throw runtime_error("invalid option " + option);
		} else {
			m_packageArchiveName = option;
		}
	}
	if (m_packageArchiveName.empty()) {
		m_actualError = OPTION_MISSING;
		treatErrors("");
	}
	if (m_root.empty())
		m_root="/";
	else
		m_root=m_root+"/";

	// Check UID
	if (getuid())
	{
		m_actualError = ONLY_ROOT_CAN_INSTALL_UPGRADE_REMOVE;
		treatErrors("");
	}

}
void Pkgadd::preRun()
{
	// Check UID
	if (getuid())
	{
		m_actualError = ONLY_ROOT_CAN_INSTALL_UPGRADE_REMOVE;
		treatErrors("");
	}
	extractAndRunPREfromPackage(m_packageArchiveName);
}
void Pkgadd::run()
{
	// Get the list of installed packages
	getListOfPackageNames(m_root);

	// Retrieving info about all the packages
	buildDatabaseWithDetailsInfos(false);

	// Reading the archiving to find a list of files
	pair<string, pkginfo_t> package = openArchivePackage(m_packageArchiveName);

	// Checking the rules
	readRulesFile();

	bool installed = checkPackageNameExist(package.first);
	if (installed && !m_upgrade)
	{
		m_actualError = PACKAGE_ALLREADY_INSTALL;
		treatErrors (package.first);
	}
	else if (!installed && m_upgrade)
	{
		m_actualError = PACKAGE_NOT_PREVIOUSLY_INSTALL;
		treatErrors(package.first);
	}

	set<string> non_install_files = applyInstallRules(package.first, package.second, m_actionRules);
	if (!m_upgrade) {
#ifndef NDEBUG
		cerr << "Run extractAndRunPREfromPackage without upgrade" << endl;
#endif
		if (m_runPrePost) preRun();
	}
	set<string> conflicting_files = getConflictsFilesList(package.first, package.second);
	if (!conflicting_files.empty()) {
		if (m_force) {
			Db_lock lock(m_root, true);
			set<string> keep_list;
			if (m_upgrade) // Don't remove files matching the rules in configuration
				keep_list = getKeepFileList(conflicting_files, m_actionRules);
			removePackageFilesRefsFromDB(conflicting_files, keep_list); // Remove unwanted conflicts
		} else {
			copy(conflicting_files.begin(), conflicting_files.end(), ostream_iterator<string>(cerr, "\n"));
			m_actualError = LISTED_FILES_ALLREADY_INSTALLED;
			treatErrors("'" + package.first + "'" +": listed file(s) already installed (use -f to ignore and overwrite)");
		}
	}

	set<string> keep_list;
	if (m_upgrade) {
		Db_lock lock(m_root, true);
		removePackageFilesRefsFromDB(package.first);	// Remove metadata about the package removed
		keep_list = getKeepFileList(package.second.files, m_actionRules);
		removePackageFiles(package.first, keep_list);

#ifndef NDEBUG
		cerr << "Run extractAndRunPREfromPackage after upgrade" << endl;
#endif
	}
	if (m_upgrade) {
		if (m_runPrePost) preRun();
	}
	{
		Db_lock lock(m_root, true);
		// Installation progressInfo of the files on the HD
		installArchivePackage(m_packageArchiveName, keep_list, non_install_files);

		// Add the metadata about the package to the DB
		moveMetaFilesPackage(package.first,package.second);

		// Add the info about the files to the DB
		addPackageFilesRefsToDB(package.first, package.second);
	}
	if (m_runPrePost) postRun();
}
void Pkgadd::postRun()
{
	if (checkFileExist(PKG_POST_INSTALL))
	{
		m_actualAction = PKG_POSTINSTALL_START;
		progressInfo();
		process postinstall(SHELL,PKG_POST_INSTALL, 0 );
		if (postinstall.executeShell()) {
			cerr << "WARNING Run post-install FAILED. continue" << endl;
		}
		m_actualAction = PKG_POSTINSTALL_END;
		progressInfo();
		removeFile(m_root,PKG_POST_INSTALL);
	}
}
void Pkgadd::printHelp() const
{
	cout << USAGE << m_utilName << " [options] <file>" << endl
	     << OPTIONS << endl
	     << "  -i, --ignore        do not execute pre/post install scripts" << endl
	     << "  -u, --upgrade       upgrade package with the same name" << endl
	     << "  -f, --force         force install, overwrite conflicting files" << endl
	     << "  -r, --root <path>   specify alternative installation root" << endl
	     << "  -v, --version       print version and exit" << endl
	     << "  -h, --help          print help and exit" << endl;
}

set<string> Pkgadd::getKeepFileList(const set<string>& files,
	const vector<rule_t>& rules)
{
	set<string> keep_list;
	vector<rule_t> found;

	getInstallRulesList(rules, UPGRADE, found);

	for (auto i : files) {
		for (vector<rule_t>::reverse_iterator j = found.rbegin(); j != found.rend(); j++) {
			if (checkRuleAppliesToFile(*j, i)) {
				if (!(*j).action)
					keep_list.insert(keep_list.end(), i);

				break;
			}
		}
	}

#ifndef NDEBUG
	cerr << "Keep list:" << endl;
	for (auto j : keep_list) cerr << "   " << j << endl;
	cerr << endl;
#endif
	return keep_list;
}

set<string> Pkgadd::applyInstallRules(const string& name,
	pkginfo_t& info,
	const vector<rule_t>& rules)
{
	// TODO: better algo(?)
	set<string> install_set;
	set<string> non_install_set;
	vector<rule_t> found;

	getInstallRulesList(rules, INSTALL, found);

	for (auto i : info.files) {
		bool install_file = true;
		for (vector<rule_t>::reverse_iterator j = found.rbegin(); j != found.rend(); j++) {
			if (checkRuleAppliesToFile(*j, i)) {
				install_file = (*j).action;
				break;
			}
		}
		if (install_file)
			install_set.insert(install_set.end(), i);
		else
			non_install_set.insert(i);
	}
	info.files.clear();
	info.files = install_set;

#ifndef NDEBUG
	cerr << "Install set:" << endl;
	for  (auto j : info.files) cerr << "   " << j << endl;
	cerr << endl;
	cerr << "Non-Install set:" << endl;
	for (auto j : non_install_set) cerr << "   " << j << endl;
	cerr << endl;
#endif

	return non_install_set;
}
void Pkgadd::getInstallRulesList(const vector<rule_t>& rules,
	rule_event_t event, 
	vector<rule_t>& found) const
{
	for (auto i : rules ) {
		if (i.event == event)
			found.push_back(i);
	}
}
// vim:set ts=2 :
